src/goto-symex/symex_dereference.cpp

```diff
@@ -191,14 +194,68 @@ exprt goto_symext::address_arithmetic(
   return result;
 }
 
+symbol_exprt
+goto_symext::cache_dereference(exprt &dereference_result, statet &state)
+{
+  auto const cache_key = [&] {
+    auto cache_key =
+      state.field_sensitivity.apply(ns, state, dereference_result, false);
+    if(auto let_expr = expr_try_dynamic_cast<let_exprt>(dereference_result))
+    {
+      let_expr->value() = state.rename<L2>(let_expr->value(), ns).get();
+    }
+    else
+    {
+      cache_key = state.rename<L2>(cache_key, ns).get();
+    }
+    return cache_key;
+  }();
+
+  if(auto cached = state.dereference_cache.lookup(cache_key))
+  {
+    return *cached;
+  }
+
+  auto const &cache_symbol = get_fresh_aux_symbol(
+    cache_key.type(),
+    "symex",
+    "dereference_cache",
+    dereference_result.source_location(),
+    ID_C,
```

Comment:
Is hard-coding the mode as `ID_C` the right thing to do here?

Reply:
I speculate that this is a choice, but if you believe something more complex should apply I'm happy to listen/discuss.

Reply:
It probably does not matter too much, but hard-coded modes are suspicious. I would expect that the same mode be used as the mode of the underlying pointer, but then looking this up could arguably be tedious. Therefore not a high priority, but I'm concerned that cleanup will never happen if it isn't done properly right away.

Reply:
Fixed this (pending push).